### PR TITLE
TANH/Sigmoid 16-bit activation functions using LUT

### PR DESCRIPTION
**Pull request by wwwind**
We think the reference functions for 16-bit activation are too complex for efficient implementation on resource constrained platforms and propose to replace the functions with a lookup table approach as follows:

First rescale the input data to fixed range of -10.7 to +10.7. Then use a 256-entry lookup table for Sigmoid followed by linear interpolation to efficiently derive the result.

The Sigmoid LUT table is used for the TANH function, because tanh(x) = 2*sigmoid(2*x) -1 and we take into account the symmetry.

The proposed reference kernel implementation also has higher accuracy than the existing one.
On the current functions we measure a difference of up to 6.3 for sigmoid and 11.7 for tanh in quantized units compared to the floating point reference implementation over the 16-bit input range (representing -8.0 to +8.0). For the implementation of this patch we see the error reduced to less than 1.5 quantized units compared to floating point reference for both tanh and sigmoid.

### CODE REVIEW

**Comment by bjacob** [comment]
can you share benchmark results showing how this compares to the current 16bit fixedpoint implementations on various ARM CPUs?

because pure arithmetic implementations scale linearly with SIMD widths and lookup tables do not, we have typically found that despite looking expensive in source code, pure arithmetic implemenations can actually perform well. As ARM CPUs SIMD width increases, they keep looking better. For example, on a Cortex-A76 CPU (e.g. Pixel4) we can issue 2 128-bit multiplications each cycle, each doing 8 16-bit fixed-point multiplications, so each fixed-point multiplication effectively only costs one sixteenth of a cycle.

as this varies greatly from one CPU to another, we don't expect one approach to consistently outperform the other (arithmetic vs LUT).  Instead, we have started to let both coexist in source code.  We still need to pick one to be used by default but at least having both in source code makes it easy to switch.  

i don't remember the latest but some the the following people might: @renjie-liu @abattery @jianlijianli

**Comment by renjie-liu** [comment]
Thanks Elena for the pr. Like Benoit pointed out, lut is not necessarily desired from the performance point of view.

Another thing is I suspect the accuracy measurement is "correct", do you have any real model to benchmark the accuracy? Like you said, your method will use range from -10.7 to 10.7, but should that based on the real scenario?

Jaesung can probably comment more, but if you want to cover more range for your scenario, probably just add one more bits to the integer parts? (where currently it's 3 bits, so represents -8 to 8) for sigmoid?

**Comment by abattery** [comment]
Hi, Elena.

Thank you for your contributions :)

For the performance, I have seen that LUT-table methods for Tanh/Sigmoid are practically faster than fixed point arithmetic based calculations in the modern mobile ARM CPUs for integer 8 bit cases. However, it's worth double-checking again for the integer 16 bit cases. Could you share the benchmark results before and after?

It seems that you are trying to increase accuracy based on the idea that it takes an absolute value since tanh function is symmetric and does LUT where the table covers little larger than 8-bit fixed point arithmetic range. In my first looking at it, it makes sense to me. However, As Renjie said, the Sigmoid case has the same expression range [-8, 8] with the current fixed point arithmetic kernels for Sigmoid. I am not sure that this approach improves Sigmoid case. Could you share your accuracy difference results for both Sigmoid and Tanh to us and how did you measure the accuracy?

Best regards,
Jaesung

**Comment by jianlijianli** [comment]
Cache aliasing also plays key role in the performance of LUT and that is really hardware dependent. Seems to be the best option is for them to "coexist", like Benoit mentioned in his comment.

**Comment by wwwind** [comment]
Hi guys,

Thank you for the review @bjacob @abattery @jianlijianli @renjie-liu  ! 

I attached the file accuracy_test.cpp that demonstrates how the accuracy was tested:
https://gist.github.com/wwwind/0b7413833cf8a7596bee084403224fa5

The test code reports the range and total sum of the error over the whole signed 16-bit integer range. Below are results of the testing.

**For sigmoid:**

tensorflow   sigmoid implementation error -6.392090 to 5.392578, error sum -32767.042969
proposed     sigmoid implementation error -1.015625 to 1.015625, error sum -0.043635

**For tanh:**

tensorflow   tanh    implementation error -11.791016 to 11.791016, error sum 0.992188
proposed     tanh    implementation error -1.476562 to 1.476562, error sum 0.992188 

Regarding the performance benchmark, we do not have such measurements at hand. When we refer to resource constrained platforms, apart from the world of CPUs, MCUs and GPUs, we also consider neural network accelerators, the dedicated HW where small look-up table is likely to be a preferred implementation path. However, even just for CPUs, for the non-optimised reference code, we expect the proposed solution to be more efficient than the gemmlowp one due to the number of operations involved in the iterative approach used by the latter.

In the proposed solution, to compute a value, we need about 14 operations. The existing implementation is based on the iterative Newton-Raphson division algorithm to find a reciprocal, which takes around 12 operations + computation of exponent, which needs Taylor approximations with 4 terms. So, it looks a lot heavier, at least for a single result computation.

We agree that, for vectorised implementation, the performance advantage may or may not shift towards the iterative approach. Therefore, it seems that the approach proposed by @bjacob, @abattery and @jianlijianli , where both implementations will co-exist, is a sensible compromise. I would suggest that we keep a more accurate, LUT-based implementation, as a default one. Should I proceed to make the corresponding changes in this PR? What would be the best way to implement the switching mechanism between the two implementations? Could you please point to an example in the code ? 
Thanks!

**Comment by renjie-liu** [comment]
Hi,

thanks a lot for the testing! and thanks a lot for the work!

I wonder should we sum the absolute value instead of the raw value? also the large sigmoid value makes me wonder it looks really look like a direct overflow: -32767.042969 really looks like -32768 plus some number (maybe we can just try with the extreme number?) or maybe just systematically shifted -0.5 for every value?

also given you're verifying the int16 value, judging from the number both have good results: +- 12/(1 << 15) for tanh

I think it's fine to keep both approach, you can refer here for prepare: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/activations.cc#L395-L422

and here for eval: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/activations.cc#L813-L824

**Comment by gbaned** [comment]
@wwwind Could you please check renjie-liu's comments and keep us posted. Thanks!

**Comment by wwwind** [comment]
Hi @renjie-liu,

Thanks a lot for your review. I pushed an updated version. Could you please take a look ?
Also, I think that this block is not relevant anymore for LUT approach https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/activations.cc#L515 Should it be removed ?

Thanks,
Elena

File: tensorflow/lite/kernels/activations_test.cc [hunk 1]
```diff
@@ -882,18 +886,20 @@ TEST_P(LogisticOpTest, SigmoidInt16) {
   const float kMax = 32767.f / 32768.f;
   QuantizedActivationsOpModel m(
       GetRegistration(), BuiltinOperator_LOGISTIC,
-      /*input=*/{TensorType_INT16, {1, 2, 4, 1}, 8 * kMin, 8 * kMax},
-      /*output=*/{TensorType_INT16, {1, 2, 4, 1}, kMin, kMax});
+      /*input=*/{TensorType_INT16, {1, 2, 6, 1}, 8 * kMin, 8 * kMax},
+      /*output=*/{TensorType_INT16, {1, 2, 6, 1}, kMin, kMax});
   m.SetInput<int16_t>({
       0, -6, 2, 4,   //
-      3, -2, 10, 1,  //
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/activations_test.cc]: why delete this line?

File: tensorflow/lite/kernels/activations.cc [hunk 2]
```diff
@@ -798,14 +936,12 @@ TfLiteStatus TanhEval(TfLiteContext* context, TfLiteNode* node) {
     case kTfLiteInt16: {
       TanhParams params;
       params.input_left_shift = data->input_left_shift;
-      if (kernel_type == kReference) {
```
Comment (renjie-liu) [inline tensorflow/lite/kernels/activations.cc]: we intended to keep the reference ones

**Comment by gbaned** [comment]
@wwwind Could you please check reviewer comments and keep us posted. Thanks!

File: tensorflow/lite/kernels/activations.cc (hunk 2 above)
Comment (akarmi) [reply tensorflow/lite/kernels/activations.cc]: It looks like we did not fully align on this. From the discussion preceding this change, we understood that we would keep the new LUT-based implementation as a default one, and assumed that "default" == "kReference".

It also appears that, if GEMMLOWP_NEON is not defined, optimized_ops::Tanh() called in the case of kFixedPointOptimize is the same as reference_ops::Tanh() that used to be called for kReference.

So, we just followed the same code pattern as is used for Int8 and UInt8 versions.

I appreciate now that later in this file kGenericOptimized version of kernel_type seems to be registered as the default one for Tanh and Logistic (in Register_TANH()).

Do you think we should change to calling different implementations for different kernel_type? I.e.
kReference -> reference_ops::Tanh()
kFixedPointOptimize -> optimized_ops::Tanh()
kGenericOptimized -> EvalUsingLookupTableTanh16Bit() (default)

**Comment by akarmi** [comment]
@renjie-liu, thanks for the review, while Elena is away, I replied where I could. Please, let us know what you think.

File: tensorflow/lite/kernels/activations_test.cc (hunk 1 above)
Comment (wwwind) [reply tensorflow/lite/kernels/activations_test.cc]: @renjie-liu this line has been modified by replacing 10 with 8, so the input number is in the given range of the input tensor and i added 4 more numbers.

File: tensorflow/lite/kernels/activations.cc (hunk 2 above)
Comment (wwwind) [reply tensorflow/lite/kernels/activations.cc]: @renjie-liu Could you please clarify how the current implementation and the suggested one using LUT should co-exist ? Should we call the suggested as a reference, because it is more accurate ? The current one looks like a good fit for the kernel kFixedPointOptimize. Which one should be when the kernel type is not specified ?
Thanks !

**Comment by akarmi** [comment]
@renjie-liu, @jdduke, following our off-line discussion, can we try to move this PR forward? As the proposed LUT-based INT16 implementation improves the accuracy, it seems reasonable to keep it for both kReference and kGenericOptimized paths. It would also make INT16 path in the corresponding Eval and Prepare calls symmetric to the current INT8 and UINT8 ones.

File: tensorflow/lite/kernels/activations.cc [hunk 3]
```diff
@@ -870,14 +1006,12 @@ TfLiteStatus SigmoidEval(TfLiteContext* context, TfLiteNode* node) {
...
-        reference_ops::Logistic(
-            params, GetTensorShape(input), GetTensorData<int16_t>(input),
-            GetTensorShape(output), GetTensorData<int16_t>(output));
-      } else {
+      if (kernel_type == kFixedPointOptimized) {
         optimized_ops::Logistic(
             params, GetTensorShape(input), GetTensorData<int16_t>(input),
             GetTensorShape(output), GetTensorData<int16_t>(output));
+      } else {
+        EvalUsingLookupTableSigmoid16Bit(data, input, output);
```
Comment (jianlijianli) [inline tensorflow/lite/kernels/activations.cc]: Just to make sure I'm understanding this correctly, this changes TFLite's builtin sigmoid ("kGenericOptimized") from optimized_ops::Logistic() to "EvalUsingLookupTableSigmoid16Bit", which is more accurate. And "EvalUsingLookupTableSigmoid16Bit" is not vectorized yet (like in "EvalUsingLookupTable").

Will "EvalUsingLookupTableSigmoid16Bit" be optimized in the future?

**Review by jianlijianli — COMMENTED** [review]
Thanks for the PR. Added a few comments for me to better understand the interaction with existing code.

File: tensorflow/lite/kernels/activations.cc (hunk 3 above)
Comment (akarmi) [reply tensorflow/lite/kernels/activations.cc]: As with other operators with 16-bit activations, our initial plans are to introduce the reference implementations. We have tentative plans for optimized implementations. Our other relevant efforts are currently focused on implementing corresponding kernels in TensorFlow Lite Micro, but that work depends on the progress with introducing the reference code to TFLite first.

File: tensorflow/lite/kernels/activations.cc [hunk 4]
```diff
@@ -509,6 +643,10 @@ TfLiteStatus SigmoidPrepare(TfLiteContext* context, TfLiteNode* node) {
       PopulateLookupTable<int8_t>(data, input, output, [](float value) {
         return 1.0f / (1.0f + std::exp(-value));
       });
+    } else if (input->type == kTfLiteInt16) {
+      TF_LITE_ENSURE(context, output->params.scale == 1. / 32768);
+      TF_LITE_ENSURE(context, output->params.zero_point == 0.);
```
[the remark this answers is not in the document]
Comment (wwwind) [reply tensorflow/lite/kernels/activations.cc]: done

File: tensorflow/lite/kernels/activations.cc [hunk 5]
```diff
@@ -60,7 +60,8 @@ struct OpData {
   int input_left_shift = 0;
   int32_t input_range_radius = 0;
   int diff_min = 0;
-  uint8_t table[256] = {0};
+  uint16_t table[256] = {0};
```
[the remark this answers is not in the document]
Comment (wwwind) [reply tensorflow/lite/kernels/activations.cc]: Yes, you are right. I did this implementation, before Neon optimization has been added.
I corrected, so it will be working properly for int8/uint8.

File: tensorflow/lite/kernels/activations.cc (hunk 3 above)
Comment (jianlijianli) [reply tensorflow/lite/kernels/activations.cc]: Thanks, I was not in some of the discussions so I might miss context here. Isn't it safer to change just the reference code and leave the two optimized unchanged?

Comment (akarmi) [reply tensorflow/lite/kernels/activations.cc]: The issue with keeping the old kGenericOptimized implementation is the fact that it is the one currently used as the default BuiltinOp (see [Register_TANH()](https://github.com/tensorflow/tensorflow/blob/38eeb4f5d18c6772886b1f41093f4681bb522108/tensorflow/lite/kernels/activations.cc#L1342)). One of the incentives for this PR is to be able to run models quantized with higher accuracy provided by 16-bit activations using the default/standard TFLite build.

I would suggest that we enable the proposed code by default, and address any performance issues in due course. The alternative approach could be using kReference implementation by default, but, since it would affect execution paths for other, potentially more widely used, data types, I think it may not be the best option.

Comment (jianlijianli) [reply tensorflow/lite/kernels/activations.cc]: Thanks Anton. I see your point and that makes a lot of sense. The only concern I have is that there are already model that runs with 16bit builtin Tanh/Sigmoid Op. This change will make it more accurate but in the same time slower, which can be an issue.

May I suggest the following? We can submit the code but only enable it in the reference case. Once the optimization is finished, we can flip the switch so existing users can get better accuracy without performance impact.

Does that sound good? Thanks!

File: tensorflow/lite/kernels/activations.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (jianlijianli) [reply tensorflow/lite/kernels/activations.cc]: Thanks for making the change.

File: tensorflow/lite/kernels/activations.cc (hunk 3 above)
Comment (akarmi) [reply tensorflow/lite/kernels/activations.cc]: Thanks Jian Li. I appreciate your concerns, and generally agree with the proposed course of action. One thing that I would like to do before making the mods you are asking for is to move the new reference implementation to the [reference_integer_ops](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/internal/reference/integer_ops/tanh.h) namespace. It would make it possible to re-use the reference implementation in the code external to the TFLite run-time.

What do you think?

Comment (jianlijianli) [reply tensorflow/lite/kernels/activations.cc]: Thanks for understanding the restrictions.

We all appreciate the improved accuracy this PR brings. What you mentioned works and please go ahead with the plan (either here or in another PR, whichever you feel comfortable with).

Thanks!

Comment (jianlijianli) [reply tensorflow/lite/kernels/activations.cc]: Thanks for understanding the restrictions.

We all appreciate the improved accuracy this PR brings. What you mentioned works and please go ahead with the plan (either here or in another PR, whichever you feel comfortable with).

Thanks!

**Comment by wwwind** [comment]
Hi @jianlijianli I moved implementation of Tanh/Sigmoid into integer_reference_ops per discussion. Could you please review this PR ?
Thanks!

**Review by jianlijianli — DISMISSED** [review]
Thanks for making the changes and for your patience going through the review process.

File: tensorflow/lite/kernels/internal/common.h [hunk 6]
```diff
@@ -195,6 +195,38 @@ inline int CountLeadingSignBits(T integer_input) {
 #endif
 }
 
+// Table of sigmoid(i/24) at 0.16 format - 256 elements.
+
+// We use combined sigmoid and tanh look-up table, since
+// tanh(x) = 2*sigmoid(2*x) -1.
+// Both functions are symmetric, so the LUT table is only needed
+// for the absolute value of the input.
+static uint16_t sigmoid_table_uint16[256] = {
```
Comment (rthadur) [inline tensorflow/lite/kernels/internal/common.h]: ./third_party/tensorflow/lite/kernels/internal/common.h:204:17: error: unused variable 'sigmoid_table_uint16' [-Werror,-Wunused-variable]
static uint16_t sigmoid_table_uint16[256] = {

**Review by rthadur — CHANGES_REQUESTED** [review]
@wwwind can you please check above error

**Comment by wwwind** [comment]
Hi @jianlijianli , Could you please re-approve this PR ? I had to push a fix for the warning.
Thanks!

**Review by jianlijianli — DISMISSED** [review]
Thanks for fixing the warning.

**Comment by wwwind** [comment]
Hi @gbaned, @liufengdb  ! I pushed a small fix for the error reported by buildifier.
Could you please re-approve this PR ?
Thanks!

**Review by rthadur — APPROVED** [review]

**Comment by wwwind** [comment]
Hi @rthadur ! This PR has been approved 10 days ago and all checks are green.
Is it stuck internally with some errors ? Could you please take a look?
Thanks!

**Comment by rthadur** [comment]
@wwwind thanks for your patience , @jianlijianli is working internally on this , we will get back to you asap.

**Review by jdduke — APPROVED** [review]
